File: src/router.c
/*
 * ipv6-relay - Lightweight IPv6 Relay Daemon
 * Copyright (C) 2026
 *
 * Based on odhcpd by Steven Barth
 * Licensed under GPLv2
 */

#include <errno.h>
#include <fcntl.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <stdbool.h>
#include <arpa/inet.h>
#include <net/route.h>
#include <netinet/ip6.h>

#include "ipv6_relay.h"

/*
 * RFC 4861 host constants for router solicitation. We (ab)use them on the
 * master (upstream-facing) relay interface: after (re)start or a link-up
 * event the daemon proactively solicits a Router Advertisement instead of
 * passively waiting for the next unsolicited one (which upstream may only
 * emit every MaxRtrAdvInterval - up to several minutes), so downstream
 * clients regain their prefix/default route within seconds.
 */
#define MAX_RTR_SOLICITATIONS		3
#define RTR_SOLICITATION_INTERVAL	4000	/* milliseconds */

static void send_router_solicitation(const struct interface *iface);
static void forward_router_solicitation(const struct interface *iface);
static void forward_router_advertisement(const struct interface *iface, uint8_t *data, size_t len);
static void solicit_timer_cb(struct uloop_timeout *t);

static void handle_icmpv6(void *addr, void *data, size_t len,
		struct interface *iface, void *dest);
static void router_netevent_cb(unsigned long event, struct netevent_handler_info *info);

static struct netevent_handler router_netevent_handler = { .cb = router_netevent_cb, };

int router_init(void)
{
	int ret = 0;

	if (netlink_add_netevent_handler(&router_netevent_handler) < 0) {
		error("Failed to add router netevent handler");
		ret = -1;
	}

	return ret;
}

int router_setup_interface(struct interface *iface, bool enable)
{
	int ret = 0;

	enable = enable && (iface->ra != MODE_DISABLED);

	if (!enable && iface->router_event.uloop.fd >= 0) {
		uloop_timeout_cancel(&iface->timer_rs);
		uloop_fd_delete(&iface->router_event.uloop);
		close(iface->router_event.uloop.fd);
		iface->router_event.uloop.fd = -1;
	} else if (enable) {
		struct icmp6_filter filt;
		struct ipv6_mreq mreq;
		int val = 2;

		if (iface->router_event.uloop.fd < 0) {
			/* Open ICMPv6 socket */
			iface->router_event.uloop.fd = socket(AF_INET6, SOCK_RAW | SOCK_CLOEXEC,
								IPPROTO_ICMPV6);
			if (iface->router_event.uloop.fd < 0) {
				error("socket(AF_INET6): %m");
				ret = -1;
				goto out;
			}

			if (setsockopt(iface->router_event.uloop.fd, SOL_SOCKET, SO_BINDTODEVICE,
						iface->ifname, strlen(iface->ifname)) < 0) {
				error("setsockopt(SO_BINDTODEVICE): %m");
				ret = -1;
				goto out;
			}

			/* Let the kernel compute our checksums */
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_RAW, IPV6_CHECKSUM,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_CHECKSUM): %m");
				ret = -1;
				goto out;
			}

			/* This is required by RFC 4861 */
			val = 255;
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_MULTICAST_HOPS,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_MULTICAST_HOPS): %m");
				ret = -1;
				goto out;
			}

			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_UNICAST_HOPS,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_UNICAST_HOPS): %m");
				ret = -1;
				goto out;
			}

			/* We need to know the source interface */
			val = 1;
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_RECVPKTINFO,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_RECVPKTINFO): %m");
				ret = -1;
				goto out;
			}

			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_RECVHOPLIMIT,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_RECVHOPLIMIT): %m");
				ret = -1;
				goto out;
			}

			/* Don't loop back */
			val = 0;
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_MULTICAST_LOOP,
						&val, sizeof(val)) < 0) {
				error("setsockopt(IPV6_MULTICAST_LOOP): %m");
				ret = -1;
				goto out;
			}

			/* Filter ICMPv6 package types */
			ICMP6_FILTER_SETBLOCKALL(&filt);
			ICMP6_FILTER_SETPASS(ND_ROUTER_ADVERT, &filt);
			ICMP6_FILTER_SETPASS(ND_ROUTER_SOLICIT, &filt);
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_ICMPV6, ICMP6_FILTER,
						&filt, sizeof(filt)) < 0) {
				error("setsockopt(ICMP6_FILTER): %m");
				ret = -1;
				goto out;
			}

			iface->router_event.handle_dgram = handle_icmpv6;
			iface->ra_sent = 0;
			odhcpd_register(&iface->router_event);
		} else {
			memset(&mreq, 0, sizeof(mreq));
			mreq.ipv6mr_interface = iface->ifindex;
			inet_pton(AF_INET6, ALL_IPV6_NODES, &mreq.ipv6mr_multiaddr);
			setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_DROP_MEMBERSHIP,
				   &mreq, sizeof(mreq));

			inet_pton(AF_INET6, ALL_IPV6_ROUTERS, &mreq.ipv6mr_multiaddr);
			setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_DROP_MEMBERSHIP,
				   &mreq, sizeof(mreq));
		}

		memset(&mreq, 0, sizeof(mreq));
		mreq.ipv6mr_interface = iface->ifindex;

		if (iface->master) {
			/* Master (upstream) interface receives unsolicited RAs,
			 * which are sent to the all-nodes multicast address. */
			inet_pton(AF_INET6, ALL_IPV6_NODES, &mreq.ipv6mr_multiaddr);
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_ADD_MEMBERSHIP,
						&mreq, sizeof(mreq)) < 0) {
				error("setsockopt(IPV6_ADD_MEMBERSHIP): %m");
				ret = -1;
				goto out;
			}

			/*
			 * Kick off proactive router solicitation so downstream
			 * clients get an RA within seconds of (re)start or link-up
			 * instead of waiting for the next unsolicited upstream RA.
			 */
			iface->ra_sent = 0;
			iface->timer_rs.cb = solicit_timer_cb;
			uloop_timeout_set(&iface->timer_rs, 100);
		} else {
			/* Slave (downstream) interface receives RSs from hosts,
			 * which are sent to the all-routers multicast address. */
			inet_pton(AF_INET6, ALL_IPV6_ROUTERS, &mreq.ipv6mr_multiaddr);
			if (setsockopt(iface->router_event.uloop.fd, IPPROTO_IPV6, IPV6_ADD_MEMBERSHIP,
						&mreq, sizeof(mreq)) < 0) {
				error("setsockopt(IPV6_ADD_MEMBERSHIP): %m");
				ret = -1;
				goto out;
			}
		}
	}

out:
	if (ret < 0 && iface->router_event.uloop.fd >= 0) {
		close(iface->router_event.uloop.fd);
		iface->router_event.uloop.fd = -1;
	}

	return ret;
}

/* Send a router solicitation out a single (master) interface */
static void send_router_solicitation(const struct interface *iface)
{
	struct sockaddr_in6 all_routers;
	struct nd_router_solicit rs = {
		.nd_rs_type = ND_ROUTER_SOLICIT,
		.nd_rs_code = 0,
		.nd_rs_cksum = 0,
		.nd_rs_reserved = 0,
	};

	if (iface->router_event.uloop.fd < 0)
		return;

	memset(&all_routers, 0, sizeof(all_routers));
	all_routers.sin6_family = AF_INET6;
	all_routers.sin6_port = htons(IPPROTO_ICMPV6);
	inet_pton(AF_INET6, ALL_IPV6_ROUTERS, &all_routers.sin6_addr);

	odhcpd_send(iface->router_event.uloop.fd, &all_routers,
			&(struct iovec){&rs, sizeof(rs)}, 1, iface);
}

/*
 * Timer callback: proactively solicit an RA from upstream on the master
 * interface. Repeated a few times (RFC 4861 MAX_RTR_SOLICITATIONS) to
 * survive a lost packet and to cover the case where the upstream link
 * only became usable shortly after we started.
 */
static void solicit_timer_cb(struct uloop_timeout *t)
{
	struct interface *iface = container_of(t, struct interface, timer_rs);

	if (!iface->master || iface->ra != MODE_RELAY ||
			iface->router_event.uloop.fd < 0)
		return;

	send_router_solicitation(iface);

	if (++iface->ra_sent < MAX_RTR_SOLICITATIONS)
		uloop_timeout_set(&iface->timer_rs, RTR_SOLICITATION_INTERVAL);
}

/* Forward router solicitation */
static void forward_router_solicitation(const struct interface *iface)
{
	struct interface *c;

	avl_for_each_element(&interfaces, c, avl) {
		if (c->ifindex == iface->ifindex || !c->master || c->ra != MODE_RELAY)
			continue;

		/* Must send on the destination interface's own socket, since
		 * each socket is bound (SO_BINDTODEVICE) to its own iface. */
		send_router_solicitation(c);
	}
}

/* Forward router advertisement */
static void forward_router_advertisement(const struct interface *iface, uint8_t *data, size_t len)
{
	struct interface *c;
	struct sockaddr_in6 all_nodes;

	memset(&all_nodes, 0, sizeof(all_nodes));
	all_nodes.sin6_family = AF_INET6;
	all_nodes.sin6_port = htons(IPPROTO_ICMPV6);
	inet_pton(AF_INET6, ALL_IPV6_NODES, &all_nodes.sin6_addr);

	avl_for_each_element(&interfaces, c, avl) {
		if (c->ifindex == iface->ifindex || c->master || c->ra != MODE_RELAY)
			continue;

		/* Must send on the destination interface's own socket, since
		 * each socket is bound (SO_BINDTODEVICE) to its own iface. */
		odhcpd_send(c->router_event.uloop.fd, &all_nodes,
				&(struct iovec){data, len}, 1, c);
	}
}

/* Handle ICMPv6 messages of the incoming socket */
static void handle_icmpv6(void *addr, void *data, size_t len,
		struct interface *iface, void *dest)
{
	(void)addr;
	(void)dest;
	/*
	 * Raw AF_INET6/IPPROTO_ICMPV6 sockets deliver only the ICMPv6
	 * payload on receive - unlike IPv4 raw sockets, the IPv6 header is
	 * never prepended (RFC 3542 Section 3). "data" therefore already
	 * starts at the ICMPv6 header.
	 */
	struct icmp6_hdr *hdr = data;

	/* Ignore if too short to contain an ICMPv6 header */
	if (len < sizeof(*hdr))
		return;

	/* For relay mode, just forward RS and RA between interfaces */
	if (iface->ra == MODE_RELAY) {
		if (hdr->icmp6_type == ND_ROUTER_SOLICIT) {
			forward_router_solicitation(iface);
		} else if (hdr->icmp6_type == ND_ROUTER_ADVERT) {
			forward_router_advertisement(iface, data, len);
		}
	}
}

static void router_netevent_cb(unsigned long event, struct netevent_handler_info *info)
{
	if (event == NETEV_ADDR6_ADD || event == NETEV_ADDR6_DEL ||
	    event == NETEV_ADDRLIST_CHANGE) {
		struct interface *i;
		avl_for_each_element(&interfaces, i, avl) {
			ssize_t cnt = netlink_get_interface_addrs(i->ifindex, true, &i->addr6);
			if (cnt >= 0)
				i->addr6_len = cnt;
		}
	}
}

File: internal/relay/router.go
package relay

import (
	"net/netip"
	"time"

	"golang.org/x/sys/unix"
)

// RFC 4861 host constants, (ab)used on the master (upstream-facing) relay
// interface: after (re)start or link-up we proactively solicit an RA
// instead of passively waiting for the next unsolicited one.
const (
	maxRtrSolicitations     = 3
	rtrSolicitationInterval = 4 * time.Second
)

const (
	ndRouterSolicit = 133
	ndRouterAdvert  = 134

	ndOptSourceLinkAddr = 1
)

type routerSock struct {
	fd    int
	iface *Interface
	done  chan struct{}
}

// routerSetup mirrors router_setup_interface(): open (or tear down) the raw
// ICMPv6 socket used to relay Router Solicitations/Advertisements between
// interfaces.
func routerSetup(iface *Interface, enable bool) error {
	enable = enable && iface.RA != ModeDisabled

	if iface.router != nil {
		close(iface.router.done)
		closeFD(iface.router.fd)
		iface.router = nil
	}

	if !enable {
		if iface.rsTimer != nil {
			iface.rsTimer.Stop()
			iface.rsTimer = nil
		}
		return nil
	}

	fd, err := unix.Socket(unix.AF_INET6, unix.SOCK_RAW|unix.SOCK_CLOEXEC, unix.IPPROTO_ICMPV6)
	if err != nil {
		Errorf("socket(AF_INET6) for router relay on %s: %v", iface.Ifname, err)
		return err
	}

	ok := false
	defer func() {
		if !ok {
			closeFD(fd)
		}
	}()

	if err := bindToDevice(fd, iface.Ifname); err != nil {
		Errorf("SO_BINDTODEVICE(%s): %v", iface.Ifname, err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_RAW, unix.IPV6_CHECKSUM, 2); err != nil {
		Errorf("IPV6_CHECKSUM: %v", err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_IPV6, unix.IPV6_MULTICAST_HOPS, 255); err != nil {
		Errorf("IPV6_MULTICAST_HOPS: %v", err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_IPV6, unix.IPV6_UNICAST_HOPS, 255); err != nil {
		Errorf("IPV6_UNICAST_HOPS: %v", err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_IPV6, unix.IPV6_RECVPKTINFO, 1); err != nil {
		Errorf("IPV6_RECVPKTINFO: %v", err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_IPV6, unix.IPV6_RECVHOPLIMIT, 1); err != nil {
		Errorf("IPV6_RECVHOPLIMIT: %v", err)
		return err
	}
	if err := setsockoptInt(fd, unix.IPPROTO_IPV6, unix.IPV6_MULTICAST_LOOP, 0); err != nil {
		Errorf("IPV6_MULTICAST_LOOP: %v", err)
		return err
	}

	filt := icmp6FilterBlockAll()
	icmp6FilterSetPass(&filt, ndRouterAdvert)
	icmp6FilterSetPass(&filt, ndRouterSolicit)
	if err := unix.SetsockoptICMPv6Filter(fd, unix.IPPROTO_ICMPV6, icmpv6FilterOpt, &filt); err != nil {
		Errorf("ICMP6_FILTER: %v", err)
		return err
	}

	group := allIPv6Routers
	if iface.Master {
		group = allIPv6Nodes
	}
	if err := joinMulticast6(fd, group, iface.Ifindex, true); err != nil {
		Errorf("IPV6_ADD_MEMBERSHIP: %v", err)
		return err
	}

	rs := &routerSock{fd: fd, iface: iface, done: make(chan struct{})}
	iface.router = rs
	iface.raSent = 0
	ok = true

	go routerReadLoop(rs)

	if iface.Master {
		iface.rsTimerN++
		gen := iface.rsTimerN
		iface.rsTimer = time.AfterFunc(100*time.Millisecond, func() {
			Eng.Post(func() { solicitTimerFire(iface, gen) })
		})
	}

	return nil
}

func solicitTimerFire(iface *Interface, gen int) {
	if iface.rsTimerN != gen || iface.router == nil || !iface.Master || iface.RA != ModeRelay {
		return
	}

	// A link-up event can precede IPv6 link-local address assignment. Sending
	// before the address exists fails with EADDRNOTAVAIL; wait without
	// consuming one of the three actual solicitation attempts.
	if _, ok := getLinkLocalAddr(iface); !ok {
		iface.rsTimer = time.AfterFunc(rtrSolicitationInterval, func() {
			Eng.Post(func() { solicitTimerFire(iface, gen) })
		})
		return
	}

	if !sendRouterSolicitation(iface) {
		iface.rsTimer = time.AfterFunc(rtrSolicitationInterval, func() {
			Eng.Post(func() { solicitTimerFire(iface, gen) })
		})
		return
	}

	iface.raSent++
	if iface.raSent < maxRtrSolicitations {
		iface.rsTimer = time.AfterFunc(rtrSolicitationInterval, func() {
			Eng.Post(func() { solicitTimerFire(iface, gen) })
		})
	}
}

// sendRouterSolicitation mirrors send_router_solicitation(): send an RS to
// the all-routers multicast address, including a Source Link-Layer Address
// option per RFC 4861 4.1 whenever we know our own MAC.
func sendRouterSolicitation(iface *Interface) bool {
	if iface.router == nil {
		return false
	}

	pkt := make([]byte, 8)
	pkt[0] = ndRouterSolicit // type
	// code=0, checksum computed by kernel, reserved=0

	if mac, err := getMAC(iface); err == nil {
		opt := make([]byte, 8)
		opt[0] = ndOptSourceLinkAddr
		opt[1] = 1 // length in units of 8 bytes
		copy(opt[2:], mac[:6])
		pkt = append(pkt, opt...)
	}

	return sendICMP6(iface, allIPv6Routers, iface.Ifindex, netip.Addr{}, pkt) == nil
}

// forwardRouterSolicitation mirrors forward_router_solicitation(): resend
// an RS out every other master interface's own socket.
func forwardRouterSolicitation(iface *Interface) {
	for _, c := range interfaces {
		if c.Ifindex == iface.Ifindex || !c.Master || c.RA != ModeRelay {
			continue
		}
		sendRouterSolicitation(c)
	}
}

// fixSourceLinkAddrOption rewrites the Source Link-Layer Address option (if
// present) of a Router Advertisement in place to carry mac, mirroring
// fix_source_linkaddr_option(). See the C comment for why this matters:
// without it, downstream hosts learn the *upstream* router's MAC for what
// is now our own (rewritten-source) RA, breaking Neighbor Unreachability
// Detection intermittently.
func fixSourceLinkAddrOption(data []byte, mac [6]byte) {
	const raHdrLen = 16 // sizeof(struct nd_router_advert)
	if len(data) < raHdrLen {
		return
	}

	opt := raHdrLen
	for opt+2 <= len(data) {
		optType := data[opt]
		optLen := int(data[opt+1]) * 8
		if optLen == 0 || opt+optLen > len(data) {
			break
		}

		if optType == ndOptSourceLinkAddr {
			addrLen := optLen - 2
			n := addrLen
			if n > 6 {
				n = 6
			}
			copy(data[opt+2:opt+2+n], mac[:n])
		}

		opt += optLen
	}
}

// forwardRouterAdvertisement mirrors forward_router_advertisement(): relay
// an RA received on the master interface to every slave interface, with a
// private copy of the packet so each slave gets the SLLA option patched to
// its own MAC.
func forwardRouterAdvertisement(iface *Interface, data []byte) {
	for _, c := range interfaces {
		if c.Ifindex == iface.Ifindex || c.Master || c.RA != ModeRelay {
			continue
		}

		out := data
		if mac, err := getMAC(c); err == nil {
			buf := make([]byte, len(data))
			copy(buf, data)
			fixSourceLinkAddrOption(buf, hwAddrBytes6(mac))
			out = buf
		}

		sendICMP6(c, allIPv6Nodes, c.Ifindex, netip.Addr{}, out)
	}
}

// sendICMP6 sends an ICMPv6 packet on iface's router socket toward dest,
// mirroring plain odhcpd_send() (no explicit source address override - the
// kernel picks one). Both send_router_solicitation() and
// forward_router_advertisement() in the C daemon use this plain variant;
// only ndp.c's relay_ping() needs the link-local-source variant (see
// sendICMP6FD in rawsock.go).
func sendICMP6(iface *Interface, dest netip.Addr, ifindex int, _ netip.Addr, payload []byte) error {
	if iface.router == nil {
		return unix.EBADF
	}
	return sendICMP6FD(iface.router.fd, iface, dest, ifindex, netip.Addr{}, payload)
}

// routerReadLoop reads ICMPv6 packets from a router socket and posts them
// to the Engine, mirroring odhcpd_receive_packets() for this socket type.
func routerReadLoop(rs *routerSock) {
	buf := make([]byte, 8192)
	oob := make([]byte, 256)

	for {
		select {
		case <-rs.done:
			return
		default:
		}

		n, oobn, _, _, err := unix.Recvmsg(rs.fd, buf, oob, 0)
		if err != nil {
			if err == unix.EINTR {
				continue
			}
			select {
			case <-rs.done:
				return
			default:
			}
			time.Sleep(10 * time.Millisecond)
			continue
		}
		if n == 0 {
			continue
		}

		_, hoplimit, hasHL := parseCmsgs(oob[:oobn])
		if hasHL && hoplimit != 255 {
			continue
		}

		data := make([]byte, n)
		copy(data, buf[:n])

		Eng.Post(func() { handleICMPv6(rs, data) })
	}
}

// handleICMPv6 mirrors handle_icmpv6(): validate and relay RS/RA between
// interfaces.
func handleICMPv6(rs *routerSock, data []byte) {
	iface := rs.iface
	if iface.router != rs || iface.RA != ModeRelay {
		return
	}
	if len(data) < 4 {
		return
	}

	icmpType := data[0]
	icmpCode := data[1]
	if icmpCode != 0 {
		return
	}

	switch icmpType {
	case ndRouterSolicit:
		forwardRouterSolicitation(iface)
	case ndRouterAdvert:
		forwardRouterAdvertisement(iface, data)
	}
}
